### PR TITLE
Accelerometer calibration helper

### PR DESCRIPTION
**Pull request by linguini1**
## Summary

This PR introduces a new application which streams UORB accelerometer data to the console in tab delimited format. This allows a user to record the console input on their host machine and feed it into host-based programs like [Magneto](https://sites.google.com/view/sailboatinstruments1/a-download-magneto-v1-2?authuser=0) and [FreeIMU](https://varesano.net/freeimu-magnetometer-and-accelerometer-calibration-gui-alpha-version-out/) or visualization programs like [this one](https://github.com/michaelwro/accelerometer-calibration).

## Impact

Allows for calibration of any uORB-based accelerometer through the serial console. This is more within the realm of what Adafruit users are used to with the Arduino serial console, and thus allows NuttX supported boards to leverage existing tools designed for those use cases.

This has an impact of adding a new application, as well as a new `sensors` directory as suggested in the mailing list discussion about sensor fusion applications.

Corresponding documentation pending here: https://github.com/apache/nuttx/pull/15782

## Testing

Testing performed using the LSM6DSO32 accelerometer on a custom flight computer board. Driver is still a WIP but will be contributed to upstream soon.

Here is some sample output produced for tab delimited data:
```console
NuttShell (NSH) NuttX-12.8.0
nsh> accel_calib sensor_accel0
Sampling frequency is 50Hz
0.003589        -0.227319       8.654883
0.013160        -0.253640       9.955389
0.015553        -0.245266       9.961371
0.027517        -0.253640       9.984103
0.025124        -0.265605       9.962567
0.032303        -0.272783       9.986496
0.025124        -0.289533       9.986496
0.017946        -0.288337       9.981710
0.009571        -0.277569       9.990085
-0.005982       -0.248855       9.962567
```

And with the delimiter as a comma (set through Kconfig):

```console
nsh> accel_calib sensor_accel0
Sampling frequency is 25Hz
-0.005982,-0.236891,8.745811
-0.035892,-0.245266,9.964960
-0.019142,-0.228516,9.96855
```

With the sampling frequency set to 1Hz through Kconfig:

```console
nsh> accel_calib sensor_accel0
Sampling frequency is 1Hz
0.035892    -0.276372    9.966157
0.010767    -0.277569    9.968550
0.016749    -0.271587    9.967353
0.035892    -0.277569    9.963764
0.026321    -0.275176    9.970942
0.033499    -0.287140    9.976925
0.029910    -0.273980    9.972138
```

It's not seen in the logs but the rate was much slower than 50Hz, so rate is indeed being changed properly.

When an invalid sensor topic is provided:

```console
nsh> accel_calib something
Failed to get metadata for something
```

When no sensor topic is provided:

```console
nsh> accel_calib 
Please provide the name of the accelerometer uORB topic as an argument
Usage: accel_calib [topic]
Example: accel_calib sensor_accel0
```

### CODE REVIEW

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)

__Fill In The Commit Message:__ This PR contains a Commit with an Empty Commit Message. Please fill in the Commit Message with the PR Summary.



This PR description mostly meets the NuttX requirements, but could be improved in a few areas:

**Strengths:**

* **Clear Summary:** The summary explains the purpose, functionality, and how it works.  The Magneto/FreeIMU context is helpful.
* **Impact Description:** The impact on the user and the addition of a new application and directory are well explained.
* **Testing Evidence:**  Provides sample output demonstrating the functionality and error handling.

**Weaknesses and Suggestions:**

* **Missing Issue References:**  If there are related NuttX issues or PRs, those should be linked.  Even if there isn't a directly related issue, it might be good practice to open one for discussion before submitting the PR.
* **Incomplete Impact Assessment:**  Several impact sections are missing details:
    * **Impact on build:** Does this require any new build dependencies or configuration options?  Be specific.
    * **Impact on hardware:** You mention a custom board.  Specify the architecture.  Will this work on other boards with accelerometers?  If it relies on a WIP driver, clarify that the driver is *not* part of this PR.
    * **Impact on documentation:** Since this is a new application, documentation is necessary.  Confirm that you've included it (e.g., in the apps/ directory's README or a separate document).
    * **Impact on security:** While this specific application might not have direct security implications, consider things like potential denial-of-service if the sensor data stream floods the system.  If truly no impact, explicitly state "NO - This application has no direct security implications."
    * **Impact on compatibility:**  Address backward compatibility (will this work on older NuttX versions?) and forward compatibility (any assumptions that might break in future NuttX versions?).
* **Incomplete Testing:**
    * **Build Host Details:** Specify your build host OS, CPU architecture, and compiler version.
    * **Target Details:** Be more specific about the target architecture (e.g., `armv7-m`). State the board configuration used (e.g., `stm32f4discovery:nsh`).
    * **Missing Comma Delimited Output:**  You mention testing with comma delimiters but didn't provide the output.  Include that.
    * **"Before Change" Logs:**  Since this is a new feature, the "before change" logs are not applicable.  State that explicitly.  Focus on the "after change" logs showing successful operation.
    * **More Comprehensive Testing:**  Consider edge cases. What happens if the sensor data is invalid?  What happens if the sensor disconnects? What happens with different baud rates?


**Example of Improved Impact Section:**

```
## Impact

* **Is new feature added? Is existing feature changed?** YES, a new application is added.
* **Impact on user (will user need to adapt to change)?** YES. Users will be able to use this new application to calibrate accelerometers via the serial console.
* **Impact on build (will build process change)?** YES.  This application needs to be added to the application build configuration.  For example, users will need to add `CONFIG_EXAMPLES_ACCEL_CALIB=y` to their defconfig file. No new external dependencies are introduced.
* **Impact on hardware (will arch(s) / board(s) / driver(s) change)?**  YES. This application requires an accelerometer driver providing data via UORB.  It was tested on a custom `armv7-m` board (details omitted as the hardware design is not yet public) with an LSM6DSO32 accelerometer. The driver for this accelerometer is a work-in-progress and will be contributed separately. This application should be compatible with any board supporting UORB and an accelerometer driver.
* **Impact on documentation (is update required / provided)?** YES. Documentation for this new application is included in the `apps/sensors/README.md` file.
* **Impact on security (any sort of implications)?** NO - This application has no direct security implications. However, excessive console output could potentially lead to performance degradation if the system is heavily loaded.
* **Impact on compatibility (backward/forward/interoperability)?** This application is designed to work with NuttX 12.8.0 and later. Backward compatibility with older versions has not been tested.  Forward compatibility should be maintained as long as the UORB API remains stable.
* **Anything else to consider?**  This application assumes the accelerometer driver publishes data in the standard UORB sensor format.
```

By addressing these weaknesses, you will create a much stronger and more compliant PR, increasing the chances of it being accepted. Remember to be concise and specific in your responses.

**Comment by acassis** [comment]
@linguini1 please fix these issues:

```
/home/runner/work/nuttx-apps/nuttx-apps/apps/sensors/accel_calib/accel_calib_main.c: error: No block comments found
Error: /home/runner/work/nuttx-apps/nuttx-apps/apps/sensors/accel_calib/accel_calib_main.c:1:1: error: Missing file header comment block
Warning: /home/runner/work/nuttx-apps/nuttx-apps/apps/sensors/accel_calib/accel_calib_main.c:1:1: warning: #include outside of 'Included Files' section
Error: /home/runner/work/nuttx-apps/nuttx-apps/apps/sensors/accel_calib/accel_calib_main.c:2:0: error: Path relative to repository other than "nuttx" must begin with the root directory
Warning: /home/runner/work/nuttx-apps/nuttx-apps/apps/sensors/accel_calib/accel_calib_main.c:2:1: warning: #include outside of 'Included Files' section
Warning: /home/runner/work/nuttx-apps/nuttx-apps/apps/sensors/accel_calib/accel_calib_main.c:3:1: warning: #include outside of 'Included Files' section
Warning: /home/runner/work/nuttx-apps/nuttx-apps/apps/sensors/accel_calib/accel_calib_main.c:4:1: warning: #include outside of 'Included Files' section
Warning: /home/runner/work/nuttx-apps/nuttx-apps/apps/sensors/accel_calib/accel_calib_main.c:5:1: warning: #include outside of 'Included Files' section
Warning: /home/runner/work/nuttx-apps/nuttx-apps/apps/sensors/accel_calib/accel_calib_main.c:6:1: warning: #include outside of 'Included Files' section
Warning: /home/runner/work/nuttx-apps/nuttx-apps/apps/sensors/accel_calib/accel_calib_main.c:8:1: warning: #include outside of 'Included Files' section
Warning: /home/runner/work/nuttx-apps/nuttx-apps/apps/sensors/accel_calib/accel_calib_main.c:11:1: warning: #define outside of 'Pre-processor Definitions' section
Error: /home/runner/work/nuttx-apps/nuttx-apps/apps/sensors/accel_calib/accel_calib_main.c:74:7: error: Missing whitespace after semicolon
Error: /home/runner/work/nuttx-apps/nuttx-apps/apps/sensors/accel_calib/accel_calib_main.c:74:8: error: Missing whitespace after semicolon
Error: /home/runner/work/nuttx-apps/nuttx-apps/apps/sensors/accel_calib/accel_calib_main.c:100:1: error: "Private/Public Functions" not found! File will not be checked
Error: Process completed with exit code 1.
```

**Comment by linguini1** [comment]
Unsure about these build errors, there appears to be some issues with `math.h` but I do not believe this is related to my changes.

File: sensors/Kconfig [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#
+# For a description of the syntax of this configuration file,
+# see the file kconfig-language.txt in the NuttX tools repository.
+#
+# This file is autogenerated, do not edit.
+#
+
+menu "Sensor Application Libraries"
+source "/home/linguini/coding/nuttx-space/nuttx-apps/sensors/accel_calib/Kconfig"
```
Comment (xiaoxiang781216) [inline sensors/Kconfig:9]: why add the generated soure file

File: sensors/accel_calib/accel_calib_main.c [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+      if (update)
+        {
+          err = orb_copy(imu_meta, imu, &data);
+          if (err)
+            {
+              fprintf(stderr, "Failed to get new data: %d\n", err);
+              continue;
+            }
+
+          printf("%f" DLM "%f" DLM "%f\n", data.x, data.y, data.z);
```
Comment (xiaoxiang781216) [inline sensors/accel_calib/accel_calib_main.c:145]: why not reuse the output from orb_listener or https://github.com/apache/nuttx-apps/tree/master/system/sensorscope, but write a python script on PC to convert the format that the 3rd party tool is expected.

File: sensors/Kconfig (hunk 1 above)
Comment (linguini1) [reply sensors/Kconfig:9]: Will remove!

File: sensors/accel_calib/accel_calib_main.c (hunk 2 above)
Comment (linguini1) [reply sensors/accel_calib/accel_calib_main.c:145]: I did consider this, however some of these applications read directly from a COM/serial port on the host device. I suppose maybe the UORB listener application could be extended if that functionality was required. Perhaps this application is not really necessary 

If I make a Python script I would likely not keep it in the NuttX apps, but external.

Comment (xiaoxiang781216) [reply sensors/accel_calib/accel_calib_main.c:145]: > I did consider this, however some of these applications read directly from a COM/serial port on the host device. I suppose maybe the UORB listener application could be extended if that functionality was required. Perhaps this application is not really necessary
> 

There are many tools on both Windows, Linux or macOS to insert a virtual serial port between the script and the application like this:
real serial port<->the script<->virtual serial port<->applicatin/algorithm

> If I make a Python script I would likely not keep it in the NuttX apps, but external.

host script could put into apps/tools folder.

**Comment by raiden00pl** [comment]
a loose idea of how to do it differently:

1. use `sensorscope` which is more general solution, but at this moment it doesn't support sensor configuration. It would be best if you could pass configuration parameters from the command line, but initially it can be hardcoded with Kconfig.

2. add an interface to `nxscope` which, instead of sending frames via serial port, prints data according to the type and channel parameters to the console.

In this way we would get a universal tool for working with sensors with the ability to stream data to the host with minimal overhead (nxscope is a pretty lightweight protocol).
I would like to do something like that someday, but if someone is willing to do it before me - that would be great :)

File: sensors/accel_calib/accel_calib_main.c (hunk 2 above)
Comment (linguini1) [reply sensors/accel_calib/accel_calib_main.c:145]: Thank you for the suggestions everyone! I will look into those existing solutions and see if they fit my use case/how I can extend them.

**Comment by linguini1** [comment]
@raiden00pl do you happen to have some documentation about using sensorscope? I checked out your repo linked from the `nxscope` docs, but neither `nxscope` or your docs seem to explain how to use the sensor scope or set it up. I made some attempts but got some unexpected results where errors are constantly printed to my screen.

```console
 -61
    ERROR: read failed -61
                          ERROR: read failed -61
                                                ERROR: read failed -61
                                                                      ERROR: read failed -61
                                                                                            ERROR: read failed -1
```

**Comment by raiden00pl** [comment]
I described an example with thingy53 here: https://railab.me/posts/2024/6/nuttx-sensorscope-demo-thingy53/ 
I tested `sensorscope` only with sensors `fetch` interface, so I guest your sensor uses polling method.

`-61` means that there is no data to read from sensor. You can try to remove `O_NONBLOCK` flag from here https://github.com/apache/nuttx-apps/blob/fb0c1e10ded2a6fb9f066b9893662cbcc86e4646/system/sensorscope/sensorscope_main.c#L344

and then remove this sleep: https://github.com/apache/nuttx-apps/blob/fb0c1e10ded2a6fb9f066b9893662cbcc86e4646/system/sensorscope/sensorscope_main.c#L157
